File: backend/builder/template_engine.py
import html

def escape_val(val):
    if val is None:
        return ""
    if isinstance(val, (int, float, bool)):
        return str(val)
    return html.escape(str(val))


def normalize_resume_data(data):
    if not isinstance(data, dict):
        return {}

    personal = data.get("personal", {})
    if not isinstance(personal, dict):
        personal = {}

    full_name = (
        personal.get("fullName")
        or personal.get("full_name")
        or personal.get("name")
        or data.get("fullName")
        or data.get("full_name")
        or data.get("name")
        or ""
    )

    title = personal.get("title") or personal.get("job_title") or data.get("title") or ""
    email = personal.get("email") or data.get("email") or ""
    phone = personal.get("phone") or personal.get("phone_cell") or personal.get("phone_home") or data.get("phone") or ""
    location = personal.get("location") or personal.get("address") or data.get("location") or ""
    linkedin = personal.get("linkedin") or personal.get("linkedin_url") or data.get("linkedin") or ""
    website = personal.get("website") or personal.get("portfolio") or personal.get("link") or personal.get("profile_url") or data.get("website") or ""

    summary = data.get("summary") or data.get("summary_statement") or ""

    # Normalize skills
    raw_skills = data.get("skills") or data.get("qualifications") or []
    skills = []
    if isinstance(raw_skills, list):
        for s in raw_skills:
            if isinstance(s, dict):
                skills.append(s.get("name") or s.get("skill") or "")
            elif isinstance(s, str) and s.strip():
                skills.append(s.strip())
    elif isinstance(raw_skills, str):
        skills = [s.strip() for s in raw_skills.split(",") if s.strip()]

    # Normalize experience
    raw_exp = data.get("experience") or data.get("work_experience") or data.get("jobs") or []
    experience = []
    if isinstance(raw_exp, list):
        for item in raw_exp:
            if isinstance(item, dict):
                role = item.get("role") or item.get("title") or item.get("job_title") or ""
                company = item.get("company") or item.get("organization") or ""
                loc = item.get("location") or ""
                start = item.get("start") or item.get("start_date") or ""
                end = item.get("end") or item.get("end_date") or ""
                current = item.get("current", False)
                if current:
                    duration = f"{start} – Present" if start else "Present"
                elif start and end:
                    duration = f"{start} – {end}"
                elif start:
                    duration = start
                else:
                    duration = item.get("duration") or ""

                desc = item.get("description") or item.get("desc") or ""
                bullets = item.get("bullets", [])
                if not bullets and desc:
                    bullets = [b.strip() for b in desc.split("\n") if b.strip()]

                if not role and not company and not desc:
                    continue

                experience.append({
                    "role": role,
                    "title": role,
                    "company": company,
                    "location": loc,
                    "start": start,
                    "end": "Present" if current else end,
                    "duration": duration,
                    "current": current,
                    "description": desc,
                    "bullets": bullets,
                })

    # Normalize education
    raw_edu = data.get("education") or data.get("degrees") or []
    education = []
    if isinstance(raw_edu, list):
        for item in raw_edu:
            if isinstance(item, dict):
                degree = item.get("degree") or ""
                school = item.get("school") or item.get("university") or item.get("institution") or ""
                loc = item.get("location") or ""
                start = item.get("start") or item.get("start_date") or ""
                end = item.get("end") or item.get("end_date") or ""
                duration = f"{start} – {end}" if start and end else (start or end or "")
                field = item.get("field") or item.get("major") or ""
                honor = item.get("honor") or ""
                details = item.get("details", [])

                if not degree and not school:
                    continue

                education.append({
                    "degree": degree,
                    "school": school,
                    "location": loc,
                    "start": start,
                    "end": end,
                    "duration": duration,
                    "field": field,
                    "honor": honor,
                    "details": details,
                })

    # Normalize projects
    raw_proj = data.get("projects") or []
    projects = []
    if isinstance(raw_proj, list):
        for item in raw_proj:
            if isinstance(item, dict):
                name = item.get("name") or item.get("title") or ""
                link = item.get("link") or item.get("url") or ""
                tech = item.get("technologies") or item.get("tech") or ""
                desc = item.get("description") or item.get("desc") or ""
                if not name and not desc:
                    continue
                projects.append({
                    "name": name,
                    "title": name,
                    "link": link,
                    "technologies": tech,
                    "description": desc,
                })

    # Normalize certifications
    raw_cert = data.get("certifications") or []
    certifications = []
    if isinstance(raw_cert, list):
        for item in raw_cert:
            if isinstance(item, dict):
                if item.get("name"):
                    certifications.append({
                        "name": item.get("name") or "",
                        "issuer": item.get("issuer") or item.get("organization") or "",
                        "date": item.get("date") or "",
                    })
            elif isinstance(item, str) and item.strip():
                certifications.append({"name": item.strip(), "issuer": "", "date": ""})

    sections = data.get("sections", {})
    if not isinstance(sections, dict):
        sections = {}

    normalized = {
        "name": full_name,
        "fullName": full_name,
        "full_name": full_name,
        "first_name": full_name.split()[0] if full_name else "",
        "last_name": " ".join(full_name.split()[1:]) if len(full_name.split()) > 1 else "",
        "title": title,
        "job_title": title,
        "email": email,
        "phone": phone,
        "phone_cell": phone,
        "phone_home": phone,
        "location": location,
        "address": location,
        "linkedin": linkedin,
        "website": website,
        "profile_url": website or linkedin or "#",
        "profile_label": website or linkedin or "",
        "summary": summary if sections.get("summary", True) else "",
        "summary_statement": summary if sections.get("summary", True) else "",
        "skills": skills if sections.get("skills", True) else [],
        "qualifications": skills if sections.get("skills", True) else [],
        "experience": experience if sections.get("experience", True) else [],
        "education": education if sections.get("education", True) else [],
        "projects": projects if sections.get("projects", True) else [],
        "certifications": certifications if sections.get("certifications", True) else [],
        "sections": sections,
    }
    return normalized


def resolve_lookup(ctx, path):
    if not path:
        return ""
    parts = path.strip().split(".")
    curr = ctx
    for p in parts:
        if isinstance(curr, dict) and p in curr:
            curr = curr[p]
        elif isinstance(curr, (list, tuple)) and p == "length":
            curr = len(curr)
        else:
            return ""
    return curr


def find_matching_close(template, tag_type, start_after_open_idx):
    depth = 1
    cursor = start_after_open_idx
    open_tag = f"{{{{#{tag_type}"
    close_tag = f"{{{{/{tag_type}}}}}"

    while cursor < len(template):
        next_open_idx = template.find(open_tag, cursor)
        next_close_idx = template.find(close_tag, cursor)

        if next_close_idx == -1:
            return -1

        if next_open_idx != -1 and next_open_idx < next_close_idx:
            depth += 1
            cursor = next_open_idx + len(open_tag)
        else:
            depth -= 1
            if depth == 0:
                return next_close_idx
            cursor = next_close_idx + len(close_tag)

    return -1


def render_html_list(items, tag="ul", item_class="", list_class=""):
    if not items:
        return ""
    cls_attr = f' class="{list_class}"' if list_class else ""
    li_cls = f' class="{item_class}"' if item_class else ""
    lis = "".join(f"<li{li_cls}>{escape_val(item)}</li>" for item in items if item)
    return f"<{tag}{cls_attr}>{lis}</{tag}>"


def render_default_experience(experience):
    if not experience:
        return ""
    out = []
    for exp in experience:
        dur = escape_val(exp.get("duration", ""))
        date_html = f'<div class="resume-entry-date">{dur}</div>' if dur else ""
        sub = escape_val(exp.get("company", ""))
        if exp.get("location"):
            sub += f' &mdash; {escape_val(exp.get("location"))}'
        
        desc = exp.get("description", "")
        desc_html = f'<div class="resume-entry-desc">{escape_val(desc)}</div>' if desc else ""
        
        out.append(f"""
        <div class="resume-entry">
            <div class="resume-entry-top">
                <div>
                    <div class="resume-entry-title">{escape_val(exp.get("role", ""))}</div>
                    <div class="resume-entry-sub">{sub}</div>
                </div>
                {date_html}
            </div>
            {desc_html}
        </div>
        """)
    return "\n".join(out)


def render_default_education(education):
    if not education:
        return ""
    out = []
    for edu in education:
        dur = escape_val(edu.get("duration", ""))
        date_html = f'<div class="resume-entry-date">{dur}</div>' if dur else ""
        sub = escape_val(edu.get("school", ""))
        if edu.get("location"):
            sub += f' &mdash; {escape_val(edu.get("location"))}'
        
        out.append(f"""
        <div class="resume-entry">
            <div class="resume-entry-top">
                <div>
                    <div class="resume-entry-title">{escape_val(edu.get("degree", ""))}</div>
                    <div class="resume-entry-sub">{sub}</div>
                </div>
                {date_html}
            </div>
        </div>
        """)
    return "\n".join(out)


def render_default_projects(projects):
    if not projects:
        return ""
    out = []
    for proj in projects:
        link = proj.get("link", "")
        link_html = f'<a class="resume-entry-date" href="{escape_val(link)}" target="_blank" rel="noopener noreferrer">{escape_val(link)}</a>' if link else ""
        desc = proj.get("description", "")
        desc_html = f'<div class="resume-entry-desc">{escape_val(desc)}</div>' if desc else ""
        tech = proj.get("technologies", "")
        tech_html = f'<div class="resume-project-tech"><strong>Technologies:</strong> {escape_val(tech)}</div>' if tech else ""

        out.append(f"""
        <div class="resume-entry">
            <div class="resume-entry-top">
                <div class="resume-entry-title">{escape_val(proj.get("name", ""))}</div>
                {link_html}
            </div>
            {desc_html}
            {tech_html}
        </div>
        """)
    return "\n".join(out)


def render_block(template, ctx):
    result = []
    cursor = 0

    while cursor < len(template):
        open_idx = template.find("{{#", cursor)
        if open_idx == -1:
            result.append(template[cursor:])
            break

        result.append(template[cursor:open_idx])

        tag_end_idx = template.find("}}", open_idx)
        if tag_end_idx == -1:
            result.append(template[open_idx:])
            break

        tag_content = template[open_idx + 3 : tag_end_idx].strip()
        space_idx = tag_content.find(" ")
        tag_type = tag_content[:space_idx].strip() if space_idx != -1 else tag_content
        expr = tag_content[space_idx + 1:].strip() if space_idx != -1 else ""

        close_start = tag_end_idx + 2
        close_end = find_matching_close(template, tag_type, close_start)

        if close_end == -1:
            cursor = tag_end_idx + 2
            continue

        inner_template = template[close_start:close_end]
        cursor = close_end + len(f"{{{{/{tag_type}}}}}")

        if tag_type == "if":
            val = resolve_lookup(ctx, expr)
            is_truthy = val and (not isinstance(val, (list, dict)) or len(val) > 0)
            if is_truthy:
                result.append(render_block(inner_template, ctx))
        elif tag_type == "unless":
            val = resolve_lookup(ctx, expr)
            is_truthy = val and (not isinstance(val, (list, dict)) or len(val) > 0)
            if not is_truthy:
                result.append(render_block(inner_template, ctx))
        elif tag_type == "each":
            items = resolve_lookup(ctx, expr)
            if isinstance(items, list):
                for item in items:
                    if isinstance(item, dict):
                        item_ctx = {**ctx, **item, "this": item}
                    else:
                        item_ctx = {**ctx, "this": item}
                    result.append(render_block(inner_template, item_ctx))

    rendered_text = "".join(result)

    # Standalone section tags
    if "{{skills}}" in rendered_text or "{{qualifications}}" in rendered_text:
        skills_html = render_html_list(ctx.get("skills", []), tag="ul", list_class="qual-list")
        rendered_text = rendered_text.replace("{{skills}}", skills_html).replace("{{qualifications}}", skills_html)

    if "{{experience}}" in rendered_text or "{{work_experience}}" in rendered_text:
        exp_html = render_default_experience(ctx.get("experience", []))
        rendered_text = rendered_text.replace("{{experience}}", exp_html).replace("{{work_experience}}", exp_html)

    if "{{education}}" in rendered_text:
        edu_html = render_default_education(ctx.get("education", []))
        rendered_text = rendered_text.replace("{{education}}", edu_html)

    if "{{projects}}" in rendered_text:
        proj_html = render_default_projects(ctx.get("projects", []))
        rendered_text = rendered_text.replace("{{projects}}", proj_html)

    # Replace scalar tags
    import re
    def scalar_replacer(match):
        k = match.group(1).strip()
        if k == "this":
            return escape_val(ctx.get("this", ""))
        val = resolve_lookup(ctx, k)
        if isinstance(val, (str, int, float, bool)):
            return escape_val(val)
        return ""

    return re.sub(r"\{\{([a-zA-Z0-9_.]+)\}\}", scalar_replacer, rendered_text)


def render_resume_template(template_html, resume_data):
    if not template_html:
        return ""
    ctx = normalize_resume_data(resume_data)
    return render_block(template_html, ctx)

File: frontend/src/components/builder/templateEngine.js
/**
 * Client-side safe resume template rendering engine.
 * Recursive block parser for {{#if}}, {{#unless}}, {{#each}} with exact tag balancing.
 * Matches backend/builder/template_engine.py logic 1:1.
 */

export function escapeHtml(str) {
    if (str === null || str === undefined) return ""
    if (typeof str === "number" || typeof str === "boolean") return String(str)
    return String(str)
        .replace(/&/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;")
        .replace(/'/g, "&#39;")
}

export function normalizeResumeData(data = {}) {
    const personal = data.personal || {}

    const fullName = personal.fullName || personal.full_name || personal.name || data.fullName || data.full_name || data.name || ""
    const title = personal.title || personal.job_title || data.title || ""
    const email = personal.email || data.email || ""
    const phone = personal.phone || personal.phone_cell || personal.phone_home || data.phone || ""
    const location = personal.location || personal.address || data.location || ""
    const linkedin = personal.linkedin || personal.linkedin_url || data.linkedin || ""
    const website = personal.website || personal.portfolio || personal.link || personal.profile_url || data.website || ""

    const summary = data.summary || data.summary_statement || ""

    // Skills
    const rawSkills = data.skills || data.qualifications || []
    let skills = []
    if (Array.isArray(rawSkills)) {
        skills = rawSkills.map((s) => (typeof s === "object" && s !== null ? s.name || s.skill || "" : String(s).trim())).filter(Boolean)
    } else if (typeof rawSkills === "string") {
        skills = rawSkills.split(",").map((s) => s.trim()).filter(Boolean)
    }

    // Experience
    const rawExp = data.experience || data.work_experience || data.jobs || []
    const experience = (Array.isArray(rawExp) ? rawExp : []).map((item) => {
        if (!item || typeof item !== "object") return null
        const role = item.role || item.title || item.job_title || ""
        const company = item.company || item.organization || ""
        const loc = item.location || ""
        const start = item.start || item.start_date || ""
        const end = item.end || item.end_date || ""
        const current = Boolean(item.current)
        let duration = item.duration || ""
        if (current) {
            duration = start ? `${start} – Present` : "Present"
        } else if (start && end) {
            duration = `${start} – ${end}`
        } else if (start) {
            duration = start
        }

        const desc = item.description || item.desc || ""
        let bullets = Array.isArray(item.bullets) ? item.bullets : []
        if (!bullets.length && desc) {
            bullets = desc.split("\n").map((b) => b.trim()).filter(Boolean)
        }

        // Only include if at least role, company, or description is present
        if (!role && !company && !desc) return null

        return {
            role,
            title: role,
            company,
            location: loc,
            start,
            end: current ? "Present" : end,
            duration,
            current,
            description: desc,
            bullets,
        }
    }).filter(Boolean)

    // Education
    const rawEdu = data.education || data.degrees || []
    const education = (Array.isArray(rawEdu) ? rawEdu : []).map((item) => {
        if (!item || typeof item !== "object") return null
        const degree = item.degree || ""
        const school = item.school || item.university || item.institution || ""
        const loc = item.location || ""
        const start = item.start || item.start_date || ""
        const end = item.end || item.end_date || ""
        const duration = item.duration || (start && end ? `${start} – ${end}` : start || end || "")
        const field = item.field || item.major || ""
        const honor = item.honor || ""
        const details = Array.isArray(item.details) ? item.details : []

        if (!degree && !school) return null

        return {
            degree,
            school,
            location: loc,
            start,
            end,
            duration,
            field,
            honor,
            details,
        }
    }).filter(Boolean)

    // Projects
    const rawProj = data.projects || []
    const projects = (Array.isArray(rawProj) ? rawProj : []).map((item) => {
        if (!item || typeof item !== "object") return null
        const name = item.name || item.title || ""
        const link = item.link || item.url || ""
        const tech = item.technologies || item.tech || ""
        const desc = item.description || item.desc || ""
        if (!name && !desc) return null
        return {
            name,
            title: name,
            link,
            technologies: tech,
            description: desc,
        }
    }).filter(Boolean)

    // Certifications
    const rawCert = data.certifications || []
    const certifications = (Array.isArray(rawCert) ? rawCert : []).map((item) => {
        if (typeof item === "string" && item.trim()) return { name: item.trim(), issuer: "", date: "" }
        if (!item || typeof item !== "object") return null
        if (!item.name) return null
        return {
            name: item.name || "",
            issuer: item.issuer || item.organization || "",
            date: item.date || "",
        }
    }).filter(Boolean)

    const sections = data.sections || {}

    return {
        name: fullName,
        fullName,
        full_name: fullName,
        first_name: fullName.split(" ")[0] || "",
        last_name: fullName.split(" ").slice(1).join(" ") || "",
        title,
        job_title: title,
        email,
        phone,
        phone_cell: phone,
        phone_home: phone,
        location,
        address: location,
        linkedin,
        website,
        profile_url: website || linkedin || "#",
        profile_label: website || linkedin || "",
        summary: sections.summary !== false ? summary : "",
        summary_statement: sections.summary !== false ? summary : "",
        skills: sections.skills !== false ? skills : [],
        qualifications: sections.skills !== false ? skills : [],
        experience: sections.experience !== false ? experience : [],
        education: sections.education !== false ? education : [],
        projects: sections.projects !== false ? projects : [],
        certifications: sections.certifications !== false ? certifications : [],
        sections,
    }
}

function resolveLookup(ctx, path) {
    if (!path) return ""
    const parts = path.trim().split(".")
    let curr = ctx
    for (const p of parts) {
        if (curr && typeof curr === "object" && p in curr) {
            curr = curr[p]
        } else {
            return ""
        }
    }
    return curr
}

/**
 * Finds the exact matching closing tag for a given block type (e.g. "if", "unless", "each").
 * Correctly accounts for nested blocks of the same type.
 */
function findMatchingClose(template, type, startAfterOpenIdx) {
    let depth = 1
    let cursor = startAfterOpenIdx
    const openTag = `{{#${type}`
    const closeTag = `{{/${type}}}`

    while (cursor < template.length) {
        const nextOpenIdx = template.indexOf(openTag, cursor)
        const nextCloseIdx = template.indexOf(closeTag, cursor)

        if (nextCloseIdx === -1) return -1

        if (nextOpenIdx !== -1 && nextOpenIdx < nextCloseIdx) {
            depth++
            cursor = nextOpenIdx + openTag.length
        } else {
            depth--
            if (depth === 0) return nextCloseIdx
            cursor = nextCloseIdx + closeTag.length
        }
    }
    return -1
}

function renderHtmlList(items, tag = "ul", listClass = "qual-list", itemClass = "") {
    if (!items || !items.length) return ""
    const lis = items.map((i) => `<li${itemClass ? ` class="${itemClass}"` : ""}>${escapeHtml(i)}</li>`).join("")
    return `<${tag}${listClass ? ` class="${listClass}"` : ""}>${lis}</${tag}>`
}

function renderDefaultExperience(experience) {
    if (!experience || !experience.length) return ""
    return experience.map((exp) => {
        const dateHtml = exp.duration ? `<div class="resume-entry-date">${escapeHtml(exp.duration)}</div>` : ""
        let sub = escapeHtml(exp.company || "")
        if (exp.location) sub += ` &mdash; ${escapeHtml(exp.location)}`
        const descHtml = exp.description ? `<div class="resume-entry-desc">${escapeHtml(exp.description)}</div>` : ""

        return `
        <div class="resume-entry">
            <div class="resume-entry-top">
                <div>
                    <div class="resume-entry-title">${escapeHtml(exp.role || "")}</div>
                    <div class="resume-entry-sub">${sub}</div>
                </div>
                ${dateHtml}
            </div>
            ${descHtml}
        </div>`
    }).join("\n")
}

function renderDefaultEducation(education) {
    if (!education || !education.length) return ""
    return education.map((edu) => {
        const dateHtml = edu.duration ? `<div class="resume-entry-date">${escapeHtml(edu.duration)}</div>` : ""
        let sub = escapeHtml(edu.school || "")
        if (edu.location) sub += ` &mdash; ${escapeHtml(edu.location)}`

        return `
        <div class="resume-entry">
            <div class="resume-entry-top">
                <div>
                    <div class="resume-entry-title">${escapeHtml(edu.degree || "")}</div>
                    <div class="resume-entry-sub">${sub}</div>
                </div>
                ${dateHtml}
            </div>
        </div>`
    }).join("\n")
}

function renderDefaultProjects(projects) {
    if (!projects || !projects.length) return ""
    return projects.map((proj) => {
        const linkHtml = proj.link ? `<a class="resume-entry-date" href="${escapeHtml(proj.link)}" target="_blank" rel="noopener noreferrer">${escapeHtml(proj.link)}</a>` : ""
        const descHtml = proj.description ? `<div class="resume-entry-desc">${escapeHtml(proj.description)}</div>` : ""
        const techHtml = proj.technologies ? `<div class="resume-project-tech"><strong>Technologies:</strong> ${escapeHtml(proj.technologies)}</div>` : ""

        return `
        <div class="resume-entry">
            <div class="resume-entry-top">
                <div class="resume-entry-title">${escapeHtml(proj.name || "")}</div>
                ${linkHtml}
            </div>
            ${descHtml}
            ${techHtml}
        </div>`
    }).join("\n")
}

function renderDefaultCertifications(certifications) {
    if (!certifications || !certifications.length) return ""
    return certifications.map((cert) => {
        const dateHtml = cert.date ? `<div class="resume-entry-date">${escapeHtml(cert.date)}</div>` : ""
        const issuerHtml = cert.issuer ? `<div class="resume-entry-sub">${escapeHtml(cert.issuer)}</div>` : ""

        return `
        <div class="resume-entry">
            <div class="resume-entry-top">
                <div>
                    <div class="resume-entry-title">${escapeHtml(cert.name || "")}</div>
                    ${issuerHtml}
                </div>
                ${dateHtml}
            </div>
        </div>`
    }).join("\n")
}

/**
 * Recursive block parser.
 * Handles nested {{#if}}, {{#unless}}, {{#each}} cleanly with zero mismatched tag leakage.
 */
function renderBlock(template, ctx) {
    let result = ""
    let cursor = 0

    while (cursor < template.length) {
        const openIdx = template.indexOf("{{#", cursor)
        if (openIdx === -1) {
            result += template.slice(cursor)
            break
        }

        result += template.slice(cursor, openIdx)

        const tagEndIdx = template.indexOf("}}", openIdx)
        if (tagEndIdx === -1) {
            result += template.slice(openIdx)
            break
        }

        const tagContent = template.slice(openIdx + 3, tagEndIdx).trim()
        const spaceIdx = tagContent.indexOf(" ")
        const type = spaceIdx !== -1 ? tagContent.slice(0, spaceIdx).trim() : tagContent
        const expr = spaceIdx !== -1 ? tagContent.slice(spaceIdx + 1).trim() : ""

        const closeStart = tagEndIdx + 2
        const closeEnd = findMatchingClose(template, type, closeStart)

        if (closeEnd === -1) {
            // No matching closing tag found, skip opening tag
            cursor = tagEndIdx + 2
            continue
        }

        const innerTemplate = template.slice(closeStart, closeEnd)
        cursor = closeEnd + `{{/${type}}}`.length

        if (type === "if") {
            const val = resolveLookup(ctx, expr)
            const isTruthy = val && (!Array.isArray(val) || val.length > 0) && (typeof val !== "object" || Object.keys(val).length > 0)
            if (isTruthy) {
                result += renderBlock(innerTemplate, ctx)
            }
        } else if (type === "unless") {
            const val = resolveLookup(ctx, expr)
            const isTruthy = val && (!Array.isArray(val) || val.length > 0) && (typeof val !== "object" || Object.keys(val).length > 0)
            if (!isTruthy) {
                result += renderBlock(innerTemplate, ctx)
            }
        } else if (type === "each") {
            const items = resolveLookup(ctx, expr)
            if (Array.isArray(items)) {
                for (const item of items) {
                    const itemCtx = typeof item === "object" && item !== null
                        ? { ...ctx, ...item, this: item }
                        : { ...ctx, this: item }
                    result += renderBlock(innerTemplate, itemCtx)
                }
            }
        }
    }

    // Standalone section shortcuts
    if (result.includes("{{skills}}") || result.includes("{{qualifications}}")) {
        const skillsHtml = renderHtmlList(ctx.skills, "ul", "qual-list")
        result = result.replace(/\{\{skills\}\}/g, skillsHtml).replace(/\{\{qualifications\}\}/g, skillsHtml)
    }

    if (result.includes("{{experience}}") || result.includes("{{work_experience}}")) {
        const expHtml = renderDefaultExperience(ctx.experience)
        result = result.replace(/\{\{experience\}\}/g, expHtml).replace(/\{\{work_experience\}\}/g, expHtml)
    }

    if (result.includes("{{education}}")) {
        const eduHtml = renderDefaultEducation(ctx.education)
        result = result.replace(/\{\{education\}\}/g, eduHtml)
    }

    if (result.includes("{{projects}}")) {
        const projHtml = renderDefaultProjects(ctx.projects)
        result = result.replace(/\{\{projects\}\}/g, projHtml)
    }

    if (result.includes("{{certifications}}")) {
        const certHtml = renderDefaultCertifications(ctx.certifications)
        result = result.replace(/\{\{certifications\}\}/g, certHtml)
    }

    // Replace scalar placeholders {{key}}
    return result.replace(/\{\{([a-zA-Z0-9_.]+)\}\}/g, (_, key) => {
        const k = key.trim()
        if (k === "this") {
            return escapeHtml(ctx.this)
        }
        const val = resolveLookup(ctx, k)
        if (typeof val === "string" || typeof val === "number" || typeof val === "boolean") {
            return escapeHtml(val)
        }
        return ""
    })
}

/**
 * Main export: Renders template HTML with resume data.
 */
export function renderResumeHtml(templateHtml, resumeData) {
    if (!templateHtml) return ""
    const ctx = normalizeResumeData(resumeData)
    return renderBlock(templateHtml, ctx)
}
